File: LeetCode/0446-arithmetic-slices-ii-subsequence/0446-arithmetic-slices-ii-subsequence.py
class Solution:
    def numberOfArithmeticSlices(self, nums: List[int]) -> int:
        n = len(nums)
        result = 0

        dp = [defaultdict(int) for _ in range(n)]

        for i in range(n):
            for j in range(i):
                diff = nums[i] - nums[j]
                cnt = dp[j][diff]
                result += cnt 

                dp[i][diff] += cnt + 1

        return result

File: LeetCode/0446-arithmetic-slices-ii-subsequence/0446-arithmetic-slices-ii-subsequence.cpp
class Solution {
public:
    int numberOfArithmeticSlices(vector<int>& nums) {
        int n = nums.size();
        int result = 0;

        vector<unordered_map<long long, int>> dp(n);

        for (int i = 1; i < n; ++i) {
            for (int j = 0; j < i; ++j) {
                long long diff = static_cast<long long>(nums[i]) - nums[j];
                int cnt = dp[j].count(diff) ? dp[j][diff] : 0;
                result += cnt;

                dp[i][diff] += cnt + 1;
            }
        }

        return result;
    }
};